neqo-http3/src/recv_message.rs

```diff
@@ -337,115 +345,93 @@ impl RecvMessage {
-                pseudo_header_map.entry(":status").or_insert(false);
-            }
-            MessageType::Request => {
-                pseudo_header_map.entry(":method").or_insert(false);
-                pseudo_header_map.entry(":scheme").or_insert(false);
-                pseudo_header_map.entry(":authority").or_insert(false);
-                pseudo_header_map.entry(":path").or_insert(false);
-            }
+    fn track_pseudo(&self, name: &str, state: &mut u8) -> Res<bool> {
+        let (pseudo, bit) = if name.starts_with(':') {
+            (
+                true,
+                match self.message_type {
+                    MessageType::Response => match name {
+                        ":status" => PSEUDO_HEADER_STATUS,
+                        _ => PSEUDO_HEADER_UNDEFINED,
+                    },
+                    MessageType::Request => match name {
+                        ":method" => PSEUDO_HEADER_METHOD,
+                        ":scheme" => PSEUDO_HEADER_SCHEME,
+                        ":authority" => PSEUDO_HEADER_AUTHORITY,
+                        ":path" => PSEUDO_HEADER_PATH,
+                        _ => PSEUDO_HEADER_UNDEFINED,
+                    },
+                },
+            )
+        } else {
+            (false, REGULAR_HEADER)
+        };
+
+        if *state & bit == 0 || !pseudo {
+            *state |= bit;
+            Ok(pseudo)
+        } else {
+            Err(Error::HttpGeneralProtocolStream)
         }
-        let mut seen_non_colon_header = false;
-        let excluded_headers = vec![
-            "connection",
-            "host",
-            "keep-alive",
-            "proxy-connection",
-            "te",
-            "transfer-encoding",
-            "upgrade",
-            "accept-encoding",
-        ];
-        let excluded_header_byte = vec![0x0, 0x10, 0x13, 0x3a];
+    }
+
+    fn headers_valid(&self, headers: &[Header]) -> Res<bool> {
         let mut method_value: Option<&String> = None;
+        let mut pseudo_state = 0;
         for header in headers {
-            let mut header_byte_iterator = header.0.bytes();
-            let is_colon_header = header.0.starts_with(':');
-            if is_colon_header {
-                if seen_non_colon_header {
-                    qdebug!([self], "headers_valid - pseudo header fields after fields");
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
+            let is_pseudo = self.track_pseudo(&header.0, &mut pseudo_state)?;
 
-                if let hash_map::Entry::Occupied(mut e) = pseudo_header_map.entry(&header.0) {
-                    if *e.get_mut() {
-                        return Err(Error::HttpGeneralProtocolStream);
-                    }
-                    *e.get_mut() = true;
-                } else {
+            let mut bytes = header.0.bytes();
+            if is_pseudo {
+                if (pseudo_state & (PSEUDO_HEADER_UNDEFINED | REGULAR_HEADER)) != 0 {
```

Comment:
This can be tracked in track_pseudo, right?

The former can just return an error (you don't need the bit then), the latter can be added ahead of the tuple in the main if statement (I'll mark it).